strawberryfields/program_utils.py

```diff
@@ -0,0 +1,393 @@
+"""Context for inputting a Program. Used to be a class attribute of :class:`Program`, moved
+here to avoid cyclic imports."""
+# todo: Avoid issues with Program contexts and threading,
+# cf. _pydecimal.py in the python standard distribution.
+
+
+
+def _convert(func):
+    r"""Decorator for converting user defined functions to a :class:`RegRefTransform`.
+
+    This allows classical processing of measured qumodes values.
+
+    Example usage:
+
+    .. code-block:: python
+
+        @convert
+        def F(x):
+            # some classical processing of x
+            return f(x)
+
+        with prog.context as q:
+            MeasureX       | q[0]
+            Dgate(F(q[0])) | q[1]
+
+    Args:
+        func (function): function to be converted to a :class:`RegRefTransform`.
+    """
+    @functools.wraps(func)
+    def wrapper(*args):
+        "Unused docstring."
+        return RegRefTransform(args, func)
+    return wrapper
+
+
+class RegRefError(IndexError):
+    """Exception raised by :class:`Program` when it encounters an invalid register reference.
+
+    E.g., trying to apply a gate to a nonexistent or deleted subsystem.
+    """
+
+class CircuitError(RuntimeError):
+    """Exception raised by :class:`Program` when it encounters an illegal
+    operation in the quantum circuit.
+
+    E.g., trying to use a measurement result before it is available.
+    """
+
+class MergeFailure(RuntimeError):
+    """Exception raised by :meth:`strawberryfields.ops.Operation.merge` when an
+    attempted merge fails.
+
+    E.g., trying to merge two gates of different families.
+    """
+
+
+class Command:
+    """Represents a quantum operation applied on specific subsystems of the register.
+
+    A Command instance is immutable once created, and can be shared between
+    several :class:`Program` instances.
+
+    Args:
+        op (~strawberryfields.ops.Operation): quantum operation to apply
+        reg (Sequence[RegRef]): Subsystems to which the operation is applied.
+            Note that the order matters here.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, op, reg):
+        # accept a single RegRef in addition to a Sequence
+        if not isinstance(reg, Sequence):
+            reg = [reg]
+
+        #: Operation: quantum operation to apply
+        self.op = op
+        #: Sequence[RegRef]: subsystems to which the operation is applied
+        self.reg = reg
+
+    def __str__(self):
+        """Print the command using Blackbird syntax."""
+        temp = str(self.op)
+        if self.op.ns == 0:
+            # op takes no subsystems as parameters, do not print anything more
+            return temp
+        return '{} | ({})'.format(temp, ", ".join([str(rr) for rr in self.reg]))
+
+    def __lt__(self, other):
+        # Needed as a tiebreaker for NetworkX lexicographical_topological_sort()
+        # due to a buggy implementation! Any order will do. Remove when NetworkX is fixed.
+        return True
```

Comment:
I'm not sure if anybody has reported this to NetworkX yet, the problem is this:
`lexicographical_topological_sort()` uses the Python `heapq` module to implement the algorithm.
The graph nodes are stored in the priority queue as (priority, node) pairs, and heapq compares them using the < operator, which works on tuples lexicographically, i.e. first it compares the first elements, if they are equal it compares the next elements etc.
Thus if two elements have the same priority (super common) it tries to compare the nodes themselves using `__lt__()` so you have to define it in order to avoid an error.

NetworkX does not require the graph nodes to have any methods defined, so this is a bug. The heapq documentation even provides a solution to this specific issue:

https://docs.python.org/3.7/library/heapq.html#priority-queue-implementation-notes

Reply:
I reported this as an issue to NetworkX.

https://github.com/networkx/networkx/issues/3493